### ISSUE TITLE
Routines page lacks paging controls which leads to high CPU usage with large number of procedures

### ISSUE THREAD

**Issue by erolatreston** [phpmyadmin/phpmyadmin#16487]
### Describe the bug

If the database contains large number of stored procedures (>5K), the `Routines` page take a very long time to load and the CPU is pegged to 100% during the load. With even larger number of stored procedures, the page will fail to load.

Paging controls should be added to the Routines page.

### To Reproduce

1. Create many stored procedures (5K+)
2. Expand database on the navigation tree.
3. Click on the `Procedures` in the tree

### Expected behavior

Routines page is displayed quickly with paging controls.

### Screenshots

![image](https://user-images.githubusercontent.com/13632094/99377097-b5dcab80-28bd-11eb-8476-d5c31f731d50.png)

### Server configuration

- Operating system: Centos
- Web server: apache2
- Database version: MySQL 5.7
- PHP version: 7.2
- phpMyAdmin version: 4.9.7

### Client configuration

- Browser: Chrome
- Operating system: MacOS

**Comment by williamdes** [comment]
Hello @erolatreston 
Thank you for reporting this issue !

**Comment by williamdes** [comment]
Hello @petrduda 
Would you be interested in looking for a fix for this one ?

**Comment by petrduda** [comment]
Hello @williamdes, I will have a look at this at the weekend.

**Comment by erolatreston** [comment]
It seems like when building the routines list PMA is retrieving not just the names of the routines, but the definition also. Thus, the complexity of the stored procedures severely affect the performance of this tab. Building the tab takes much longer if the stored procedures have a complex/large definition.

Good solution for this page is to change the building of this page to only use the procedures' name.

**Comment by williamdes** [comment]
Hi @erolatreston 

I found the great tool `xhprof` (quite rusty but the pecl module is working). Let me know if you could install this module on your server and the I provide you some changes to do in order to provide me some data back ?

Example for a DB list refresh:
![image](https://user-images.githubusercontent.com/7784660/105649307-04a35480-5eb0-11eb-8f13-7149607688bf.png)

**Comment by rlittle316** [comment]
@williamdes I am seeing this same issue and I am able to install xhprof on my server. I realize this has been a few years, but we would like to make progress on this issue if available. What changes should I make to properly enable the profiler to provide you some data?

**Comment by rlittle316** [comment]
Hi @williamdes I have installed php-xhprof with yum on our server, and made the changes outliined in that document, but I am still not seeing any build directory anywhere. I do not see any messages in the error log, so I am not sure what is going wrong. Is there a good way to troubleshoot this?

**Comment by rlittle316** [comment]
@williamdes I did find a way to get the profiler to work now. Once enabled do I just need to reproduce the issue and the profiles will include everything you need?

**Comment by rlittle316** [comment]
@williamdes I have been able to reproduce the issue by adding a large number of basic stored procedures with alot of text in the procedure bodies. I am still trying to figure out the best way to provide the tarball of the profiles to you. We do not have any type of file share that I am aware of.

**Comment by rlittle316** [comment]
@williamdes I have added the tarball to dropbox. Please let me know if this is sufficient for troubleshooting.

Edit by @williamdes : I used the GitHub storage for the files 

[profiles.tar.gz](https://github.com/user-attachments/files/25619905/profiles.tar.gz)

**Comment by williamdes** [comment]
Thank you @rlittle316 
I will look into the profiles

**Comment by MauricioFauth** [comment]
I merged a pull request into the `master` branch that adds a pagination to the routines page.
I'll not close this issue yet, so the performance issues could be investigated.

**Comment by kamil-tekiela** [comment]
The page is very slow because for each Routine, it needs to parse it and figure out if the routine uses any parameters. We could probably solve this problem by making the UX identical in both cases, i.e. show a dialog even if the routine doesn't take any parameters. That would not only improve the performance but also improve UX for the users.

**Comment by MauricioFauth** [comment]
Thank you, Kamil.

**Comment by MauricioFauth** [comment]
Maybe it could get the routines only from `information_schema`, this allows for better control of the result set. What do you think @kamil-tekiela?

**Comment by kamil-tekiela** [comment]
> Maybe it could get the routines only from `information_schema`, this allows for better control of the result set. What do you think [@kamil-tekiela](https://github.com/kamil-tekiela)?

I am not sure I follow you. At the moment, it can get it from IS if `DisableIS` is off. We can remove this and make it read only from IS but I don't think it would change anything. Is this what you are talking about?

**Comment by MauricioFauth** [comment]
> We can remove this and make it read only from IS but I don't think it would change anything. Is this what you are talking about?

Yes, but I mean adding a LIMIT to the query as well. And not getting all routines and then slicing the results with PHP.
